File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char>cache;
        int n = s.size();
        int j = 0;
        
        int maxLen = 0;
        for(int i=0;i<n;i++)
        {
            while(cache.find(s[i]) != cache.end())
            {
               cache.erase(s[j]);
               j++;
            }
            cache.insert(s[i]);
            maxLen = max(maxLen,(i-j+1));
        }
        return maxLen;
    }
    
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        hash_map = {}
        max_len = 0
        left = 0
        n = len(s)
        for right in range(n):
            if s[right] in hash_map and hash_map[s[right]] >= left:
                left = hash_map[s[right]] + 1
            hash_map[s[right]] = right
            max_len = max(max_len,right-left+1)
        return max_len